### PR TITLE
BUG: Fix creating a np.matrix from string syntax involving booleans

### PR DESCRIPTION
**Pull request by mirca**
This is an attempt to solve #8459.

First time contributor, appreciate any advice.

Cheers.

### CODE REVIEW

**Comment by eric-wieser** [comment]
Shouldn't this just be using `ast.literal_eval`?

**Comment by mirca** [comment]
@eric-wieser looks like we need to check for ``ValueError`` instead of ``TypeError`` after the change.

**Comment by eric-wieser** [comment]
The original code would just throw random errors anyway, since `eval` could throw all sorts of things.

So yeah, I'd say fix the tests to catch ValueError instead - ValueError is the right type of error for a misformatted string.

**Comment by eric-wieser** [comment]
Your other traceback is more problematic:

```
Traceback (most recent call last):
  File "C:\Py\lib\site-packages\numpy\matrixlib\tests\test_regression.py", line 11, in test_kron_matrix
    x = np.matrix('[1 0; 1 0]')
  File "C:\Py\lib\site-packages\numpy\matrixlib\defmatrix.py", line 230, in __new__
    data = _convert_from_string(data)
  File "C:\Py\lib\site-packages\numpy\matrixlib\defmatrix.py", line 20, in _convert_from_string
    newrow.extend(map(ast.literal_eval, temp))
  File "C:\Py\lib\ast.py", line 46, in literal_eval
    node_or_string = parse(node_or_string, mode='eval')
  File "C:\Py\lib\ast.py", line 35, in parse
    return compile(source, filename, mode, PyCF_ONLY_AST)
  File "<unknown>", line 1
    [1
     ^
SyntaxError: unexpected EOF while parsing
```

Looks like the old technique just happened to work because it discarded a bunch of random characters, including the brackets

**Comment by mirca** [comment]
@eric-wieser should it check for "(), {}" as well?

**Comment by eric-wieser** [comment]
Nah, just `[]`, since they're what matlab uses for matrices. I think everything else gets stripped out

**Comment by mirca** [comment]
@eric-wieser all green :)! Let me know if anything else is needed. Thanks for pointing out the ``ast`` module!

**Comment by eric-wieser** [comment]
I might be inclined to replace `newrow.extend(map(ast.literal_eval, temp))` with `newrow.extend([  ast.literal_eval(x.strip('[]')) for x in temp ])`, in case for some reason brackets are wanted within an item. But I guess the old code didn't allow that, so its probably best just as you have it right now.

You'll need to wait on a collaborator before this gets merged. They'll probably ask you to squash your commits, since the last two don't conform to the commit message template

**Review by eric-wieser — APPROVED** [review]

**Comment by eric-wieser** [comment]
@charris: does this need a release note?

**Comment by mirca** [comment]
Hi @eric-wieser, just to make sure here, under which milestone should I place a release note? Thanks!

**Comment by eric-wieser** [comment]
Put it in 1.13.0. In the unlikely event this is deemed worthy of a backport, it'll be copied to the 1.12.1 release notes by someone else

**Comment by mirca** [comment]
@eric-wieser Travis failures look orthogonal

**Comment by mirca** [comment]
@eric-wieser All green again. Let me know if anything else is needed, and thanks again ;-)

File: doc/release/1.13.0-notes.rst
Comment (eric-wieser) [inline doc/release/1.13.0-notes.rst]: No need to mention the details of why it failed here, or what the fix was. Only the changes the end user sees matter

**Review by eric-wieser — CHANGES_REQUESTED** [review]

File: doc/release/1.13.0-notes.rst
Comment (eric-wieser) [inline doc/release/1.13.0-notes.rst]: `np.matrix` probably wants backticks too, while you're fixing that (doesn't matter too much though, because the release notes get a final formatting pass anyway)

File: doc/release/1.13.0-notes.rst
Comment (eric-wieser) [inline doc/release/1.13.0-notes.rst]: This isn't really true. `np.matrix('(')` fails with a `SyntaxError`, and `np.matrix('{{}}')` fails with a `TypeError`.

The previous errors were also pretty random

Comment (mirca) [inline doc/release/1.13.0-notes.rst]: That's true, thanks for pointing that out!

Comment (eric-wieser) [inline doc/release/1.13.0-notes.rst]: So I'd recommend just not commenting on the exception at all.

Comment (mirca) [inline doc/release/1.13.0-notes.rst]: Sure, done.

**Comment by eric-wieser** [comment]
Dammit, made a mess of that commit message. The "merge and squash" box has more text in it than fits in the UI, and I missed the last bit

**Comment by eric-wieser** [comment]
Thanks @mirca

**Comment by KeithYJohnson** [comment]
thanks @mirca and @eric-wieser!

**Comment by mirca** [comment]
@eric-wieser hahaha, thanks :)

**Pull request merged**
